### PR TITLE
reference #207 - Filter out plugins already installed

### PR DESCRIPTION
**Pull request by jimzucker**
I updated the script to be based on jenkins.war.  I crated a local image, then created a test Dockerfile based on the image.

Please review the approach, sample output from the test run:

Removing intermediate container 7f545b034f1a
Step 7 : RUN /usr/local/bin/plugins.sh /usr/share/jenkins/ref/plugins.txt
 ---> Running in f1606faa6f4a
  ... skipping already installed:  ant:1.2
  ... skipping already installed:  antisamy-markup-formatter:1.1
  ... skipping already installed:  credentials:1.18
Downloading cucumber-testresult-plugin:0.8.2
  ... skipping already installed:  cvs:2.11
  ... skipping already installed:  external-monitor-job:1.4
Downloading git:2.4.0
Downloading git-client:1.19.0
  ... skipping already installed:  javadoc:1.1
  ... skipping already installed:  junit:1.2-beta-4
  ... skipping already installed:  ldap:1.11
  ... skipping already installed:  mailer:1.11
  ... skipping already installed:  matrix-auth:1.1
  ... skipping already installed:  matrix-project:1.4.1
  ... skipping already installed:  maven-plugin:2.7.1
  ... skipping already installed:  pam-auth:1.1
Downloading scm-api:1.0
  ... skipping already installed:  script-security:1.13
  ... skipping already installed:  ssh-credentials:1.10
  ... skipping already installed:  ssh-slaves:1.9
  ... skipping already installed:  subversion:1.54
  ... skipping already installed:  translation:1.10
  ... skipping already installed:  windows-slaves:1.0
 ---> 8da9a950147d
Removing intermediate container f1606faa6f4a
Step 8 : ADD jobs /var/jenkins_home/jobs

### CODE REVIEW

**Comment by carlossg** [comment]
Hi Jim, I'm still unclear on what you are trying to achieve.
The jenkins home will not be available at build time, just runtime, so you can't filter what plugins are already installed until runtime, which means you would have to pack them in the image anyway. The jenkins.sh could be made a bit smarter about not downgrading plugins but copying plugins that are already installed in the same version is no problem.

I see here that you are trying to avoid copying plugins already in the war into the image, but that's a corner use case, because there aren't that many in the war and basically if you don't want to copy them, just don't include them in plugins.txt

**Comment by jimzucker** [comment]
@carlossg - your right you can manually exclude them, if you understand that they are in the war.  What i experienced was my builds where much longer to iterate than they should be as most of the plugins in my instance where in the war,so by filtering I cut most of them out without having to understand any of this and my build are much shorter.

We could just update the docs more to explain all of this,but it seemed making it a built in behavior would make it a better experience for everyone so you just extract a list off plugins from the instance you want to create  container for and pass it to the script, it is smart enough to do the min work.

This is not a bug its a feature, if the team feels you dont want to go in this direction then I suggest we enhance the readme.  Let me know what you think..

**Comment by gmacario** [comment]
@carlossg, @jimzucker,

I would think that this PR can be useful not just at build time, but also when you want to update/install new Jenkins plugins inside a *running* container, i.e.

```
$ docker run -ti <running-jenkins-container> /usr/local/bin/plugin.sh some-more-plugins.txt
```

**Comment by jimzucker** [comment]
@gmacario That is a good point, I can modify it to detect when it is running and if the container is running based the decision on $JENKINS_HOME instead of inspecting the war and I would have to also make some changes to ensure it doesn't break if running multiple times concurrently.  I can make these changes but before I do can we agree if the project will want to merge it as a feature before I invest the time.  How do we reach agreement on that is unclear to me.

**Comment by jimzucker** [comment]
@gmacario, @carlossg I have updated the script  to handle the use case of running via exec, it gives a warning telling you that you must restart the container to take effect.

**Comment by jimzucker** [comment]
@gmacario  @carlossg do you need me to make further changes to move this forward?

**Comment by gmacario** [comment]
> @gmacario @carlossg do you need me to make further changes to move this forward?

@jimzucker: I have no write access to https://github.com/jenkinsci/docker so someone else should perform the merge.

However I have been testing your PR on a test branch `test-jenkinsci-docker-pull-219` of my project https://github.com/gmacario/easy-jenkins and I got an error: 

```
gmacario@ITM-GMACARIO-W7 MINGW64 ~/easy-jenkins (test-jenkinsci-docker-pull-219)
$ ./runme.sh
No active host found
INFO: Docker machine easy-jenkins exists, skipping docker-machine create
INFO: Using Docker machine easy-jenkins
Building myjenkins
Step 1 : FROM gmacario/jenkins:test-pull-219
 ---> 5627b3e3ed9f
...
Step 9 : COPY plugins.txt /usr/share/jenkins/ref/
 ---> Using cache
 ---> 56bd89b2118b
Step 10 : RUN dos2unix /usr/share/jenkins/ref/plugins.txt
 ---> Using cache
 ---> a94ed942a7b3
Step 11 : RUN /usr/local/bin/plugins.sh /usr/share/jenkins/ref/plugins.txt
 ---> Running in c941981ae81f
ERROR file not found: /usr/share/jenkins/jenkins.war
Service 'myjenkins' failed to build: The command '/bin/sh -c /usr/local/bin/plugins.sh /usr/share/jenkins/ref/plugins.txt' returned a non-zero code: 1

gmacario@ITM-GMACARIO-W7 MINGW64 ~/easy-jenkins (test-jenkinsci-docker-pull-219)
$
```

Notice that if I edit `myjenkins/Dockerfile` and rebase `FROM jenkins:latest` the script runs without problems.

Please see https://gist.github.com/gmacario/0cbd141f9e6c10a40b70840723f50e9a for details.

File: plugins.sh [hunk 1]
```diff
@@ -7,22 +7,111 @@
...
+# the plugins so we dont install 2x - there about 17!
+TEMP_ALREADY_INSTALLED=/tmp/preinstalled.plugins.$$.txt
+JENKINS_PLUGINS_DIR=/var/jenkins_home/plugins
+if [ -d $JENKINS_PLUGINS_DIR ]
+then
+	echo "Analyzing: $JENKINS_PLUGINS_DIR"
+	for i in `ls -pd1 $JENKINS_PLUGINS_DIR/*|egrep '\/$'`
+	do
+		JENKINS_PLUGIN=`basename $i`
+		JENKINS_PLUGIN_VER=`egrep -i Plugin-Version "$i/META-INF/MANIFEST.MF"|cut -d\: -f2|sed 's/ //'`
```
Comment (Vlatombe) [inline plugins.sh]: Could be great to extract this logic into a function.

**Comment by Vlatombe** [comment]
I think the idea is interesting especially if we add support for transitive dependencies in the future (#220). I'm not convinced the use case around exec is useful (in any case you'll need to restart the container, so why not just run the new image with the new plugins directly?).

File: plugins.sh [hunk 2]
```diff
@@ -7,22 +7,111 @@
...
+	if [ ! -f $JENKINS_INPUT_JOB_LIST ]
+	then
+		echo "ERROR File not found: $JENKINS_INPUT_JOB_LIST"
+		exit 1
+	fi
+fi
+
+# the war includes a # of plugins, to make the build efficient filter out 
+# the plugins so we dont install 2x - there about 17!
+TEMP_ALREADY_INSTALLED=/tmp/preinstalled.plugins.$$.txt
```
Comment (carlossg) [inline plugins.sh]: there's no guarantee that /tmp is writable, should use `JENKINS_HOME`

File: plugins.sh [hunk 3]
```diff
@@ -7,22 +7,111 @@
...
+	echo "Analyzing: $JENKINS_PLUGINS_DIR"
+	for i in `ls -pd1 $JENKINS_PLUGINS_DIR/*|egrep '\/$'`
+	do
+		JENKINS_PLUGIN=`basename $i`
+		JENKINS_PLUGIN_VER=`egrep -i Plugin-Version "$i/META-INF/MANIFEST.MF"|cut -d\: -f2|sed 's/ //'`
+		echo "$JENKINS_PLUGIN:$JENKINS_PLUGIN_VER"
+	done > $TEMP_ALREADY_INSTALLED
+else
+	JENKINS_WAR=/usr/share/jenkins/jenkins.war
+	if [ -f $JENKINS_PLUGINS_DIR ]
```
Comment (carlossg) [inline plugins.sh]: should be JENKINS_WAR

File: plugins.sh (hunk 2 above)
Comment (jimzucker) [inline plugins.sh]: OK, Changing

File: plugins.sh (hunk 3 above)
Comment (jimzucker) [inline plugins.sh]: good catch, fixed

File: plugins.sh (hunk 1 above)
Comment (jimzucker) [inline plugins.sh]: @Vlatombe i took a look at this, yes there are 2 loops but the code is pretty different so I did not do this one.  I will test and commit latter today.

**Comment by jimzucker** [comment]
@carlossg @Vlatombe  I updated based on the comments and squished the commits.  FYI I also rebased to test with the current image, it all seems OK.

**Comment by jimzucker** [comment]
@carlossg I addressed the feedback you gave on this, is there something further I need to do?

**Comment by jimzucker** [comment]
@carlossg @Vlatombe Checking in are you going to accept or reject this?  Just curious.

**Comment by carlossg** [comment]
sorry for the delay, I tried the tests and are broken, always passing for plugin installation, so I'd like to fix that before merging

**Comment by carlossg** [comment]
merged, thanks

**Pull request closed without merging**